### ISSUE TITLE
move to policy gradient approach?

### ISSUE THREAD

**Issue by mobeets** [mobeets/q-rnn#22]
Remember that one benefit of policy gradient over Q-learning is that it can learn a stochastic policy. and then we don't have to finetune the exploration during training.

note that if we use softmax exploration with Q learning, we're basically ignoring that when we train, because we're updating Q(s,a) and not the probabilities of Q(s,a). [also instead of taking max a Q(s(t+1), a), we should probably be using the softmax here as well.]

so i feel like it could be good to try to do A3C with an RNN, which rllib supports. the question is whether adding a KL penalty would be straightforward or not. see [code](https://github.com/ray-project/ray/blob/master/rllib_contrib/a3c/src/rllib_a3c/a3c/a3c_torch_policy.py) here. the question is how we would pass in the current marginal policy, since this is part of the loss. (i think we would add something to postprocess_trajectory, in the code file above)

**Comment by mobeets** [comment]
though of course, we would still have the problem that we probably don't want the KL penalty applied when taking a null action (see #20)

**Comment by mobeets** [comment]
i read this one post about how in RL, often it's better/easier to have a single-file implementation of a given algorithm, rather than using a full-on generalized library, because it's then very hard to understand how it works! i think this is a great point, and especially relevant given that we want to make a few atypical changes:
- don't encode previous action when feeding to the RNN if it's the null action (optional, but probably good to do)
- penalize deviation from the current marginal policy, not just a uniform policy (i.e., min entropy)

anyway, one option is to use [pomdp-baselines](https://github.com/twni2016/pomdp-baselines), that Ni/Eysenbach paper on using RNNs in POMDPs. they implement TD3 (which is a deterministic policy, so we don't want that), and SAC. SAC was originally implemented for continuous action spaces only, but there is SAC-D, which is what we want (and they have that implemented as well).

overall, it would take some work, but this seems like a reasonable repo to start with. check their example.ipynb file for reference.

**Comment by mobeets** [comment]
## to do

I think one useful first step could be to just take a simple, fixed representation (like, a belief representation), and add in a KL penalty, and use lucy's code to apply a policy compression type learning. then we can see in general if this approach is going to achieve what we think it is
